tests/vxlan/test_vnet_route_leak.py

```diff
@@ -0,0 +1,180 @@
+
+        if matched:
+            bgp_neighbors.append(str(matched.group(0)))
+
+    if not bgp_neighbors:
+        return False
+
+    for neighbor in bgp_neighbors:
+        if neighbor.split()[9] != "6400":
+            return False
+
+    logger.info("BGP sessions up")
+
+    return True
+
+
+def no_vnet_route_leak(duthost):
+    """
+    Gets all VNET routes and checks that they are not advertised to any BGP neighbors
+
+    Args:
+        duthost: DUT host object
+    """
+
+    vnet_routes = duthost.shell(SHOW_VNET_ROUTES_CMD)["stdout"].split("\n")
+    logger.debug("VNET prefixes: {}".format(vnet_routes))
+
+    vnet_prefixes = []
+
+    for line in vnet_routes:
+        if any(char.isdigit() for char in line):
+            vnet_prefixes.append(line.split()[1])
+
+    bgp_summary = duthost.shell(SHOW_BGP_SUMMARY_CMD)["stdout"].split("\n")
+    logger.debug("BGP Summary: {}".format(bgp_summary))
+
+    # Match only IP addresses at the beginning of the line
+    # Only IP addresses of neighbors should be matched by this
+    bgp_neighbor_addr_regex = re.compile(r"^([0-9]{1,3}\.){3}[0-9]{1,3}")
+
+    bgp_neighbors = []
+
+    for line in bgp_summary:
+        matched = bgp_neighbor_addr_regex.match(line)
+
+        if matched:
+            bgp_neighbors.append(str(matched.group(0)))
+
+    leaked_routes = {}
+
+    for neighbor in bgp_neighbors:
+        adv_routes = duthost.shell(SHOW_BGP_ADV_ROUTES_CMD_TEMPLATE.format(neighbor))["stdout"]
+
+        for prefix in vnet_prefixes:
+            if prefix in adv_routes:
+                if neighbor not in leaked_routes:
+                    leaked_routes[neighbor] = []
+                leaked_routes[neighbor].append(str(prefix))
+
+    for neighbor, routes in leaked_routes.items():
+        logger.error("VNET routes leaked to neighbor {}: {}".format(neighbor, routes))
+
+    return not leaked_routes
+
+
+def test_vnet_route_leak(configure_dut, duthost):
+    """
+    Test case for VNET route leak check
+
+    Gets a list of all VNET routes programmed to the DUT, and a list of all BGP neighbors
+    Verifies that no VNET routes are being advertised to BGP neighbors
+
+    Args:
+        configure_dut: Pytest fixture to prepare DUT for testing
+        duthost: DUT host object
+    """
+
+    assert no_vnet_route_leak(duthost)
+
+    logger.info("Restarting BGP")
+    duthost.shell(RESTART_BGP_CMD)
+
+    if not wait_until(BGP_WAIT_TIMEOUT, 5, bgp_connected, duthost):
+        logger.error("BGP sessions not established after {} seconds".format(BGP_WAIT_TIMEOUT))
+        assert False
+
+    assert no_vnet_route_leak(duthost)
+
+    logger.info("Saving and reloading CONFIG_DB")
+    duthost.shell(CONFIG_SAVE_CMD)
+    duthost.shell(CONFIG_RELOAD_CMD)
+
+    if not wait_until(BGP_WAIT_TIMEOUT, 5, bgp_connected, duthost):
+        logger.error("BGP sessions not established after {} seconds".format(BGP_WAIT_TIMEOUT))
+        assert False
+
+    assert no_vnet_route_leak(duthost)
```

Comment:
For all of these, use `pytest_assert` from the `common` package rather than bare `assert`. `assert` can lead to some weird, undescriptive error messages because of the way pytest is re-writing the asserts, and `pytest_assert` should also simplify some of these `if` checks down.

Reply:
will fix in next push